modules/mono/mono_gd/gd_mono.cpp

```diff
@@ -501,26 +501,51 @@ bool GDMono::_load_project_assembly() {
 #endif
 
 #ifdef GD_MONO_HOT_RELOAD
+void GDMono::reload_failure() {
+	if (++project_load_failure_count >= 5) {
+		// After reloading a project has failed 5 times in a row, update the path and modification time
+		// to stop any further attempts at loading this assembly, which probably is never going to work anyways
+		project_load_failure_count = 0;
+
+		ERR_PRINT_ED(".NET: Giving up on assembly reloading. Please restart the editor if unloading was failing.");
+
+		String assembly_name = GLOBAL_GET("dotnet/project/assembly_name");
+		if (assembly_name.is_empty()) {
+			assembly_name = ProjectSettings::get_singleton()->get_safe_project_name();
+		}
+		String assembly_path = GodotSharpDirs::get_res_temp_assemblies_dir().path_join(assembly_name + ".dll");
+		assembly_path = ProjectSettings::get_singleton()->globalize_path(assembly_path);
+		project_assembly_path = assembly_path.simplify_path();
+		project_assembly_modified_time = FileAccess::get_modified_time(assembly_path);
+	}
+}
+
 Error GDMono::reload_project_assemblies() {
 	ERR_FAIL_COND_V(!runtime_initialized, ERR_BUG);
 
 	finalizing_scripts_domain = true;
 
-	CSharpLanguage::get_singleton()->_on_scripts_domain_about_to_unload();
-
 	if (!get_plugin_callbacks().UnloadProjectPluginCallback()) {
-		ERR_FAIL_V_MSG(Error::FAILED, ".NET: Failed to unload assemblies.");
+		ERR_PRINT_ED(".NET: Failed to unload assemblies.");
+		reload_failure();
+		return FAILED;
 	}
 
 	finalizing_scripts_domain = false;
 
 	// Load the project's main assembly. Here, during hot-reloading, we do
 	// consider failing to load the project's main assembly to be an error.
 	if (!_load_project_assembly()) {
-		print_error(".NET: Failed to load project assembly.");
+		ERR_PRINT_ED(".NET: Failed to load project assembly.");
+		reload_failure();
 		return ERR_CANT_OPEN;
 	}
 
+	if (project_load_failure_count > 0) {
+		project_load_failure_count = 0;
+		ERR_PRINT_ED(".NET: Assembly reloading succeeded after failures.");
```

Comment:
If it was successful, should we really show an error message?

Reply:
This only appears if the unloading has failed at least once before and thus an error popup about unloading failing has already been shown to the user, this message is only to let them know that the previous error no longer matters. It could probably be an success or info message instead, except that the first doesn't exist and the second doesn't have a convenient macro for spawning one.

Reply:
Ah I see, it makes sense then.